src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java

```diff
@@ -198,18 +192,30 @@ public boolean makeDirectory(final String directory) {
     }
 
     public void transferFile(final BapSshTransfer bapSshTransfer, final FilePath filePath,
-                             final InputStream inputStream) throws SftpException {
-
-        if( !isAvoidSameFileUpload() || remoteResourceCache.checkCachedResource(filePath) ) {
-          buildInfo.printIfVerbose(Messages.console_put(filePath.getName()));
-          sftp.put(inputStream, filePath.getName());
-          success();
-        }
-        else
-        {
-          buildInfo.println(Messages._console_warning(
-            Messages.console_message_transferskip(filePath.getName()) )
-            .toString());
+                             final InputStream inputStream) throws SftpException, IOException, InterruptedException {
+        String fileName = filePath.getName();
+        buildInfo.printIfVerbose(Messages.console_put(fileName));
+        sftp.put(inputStream, fileName);
+
+        if (bapSshTransfer.isKeepFilePermissions()) {
+          FilePath parentFle = filePath.getParent();
+          if (parentFle != null) {
+            FilePath directory = parentFle.absolutize();
+            String remoteDir = sftp.pwd();
```

Comment:
Variable declaration is outside the minimal scope, where it is used. According to best practices it's better to declare variable in the closest scope

Reply:
Without declare ` FilePath parentFle = filePath.getParent()` before null check, spotbug fails with `NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE`